azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageRampupDaoImpl.java

```diff
@@ -0,0 +1,491 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package azkaban.imagemgmt.daos;
+
+import azkaban.db.DatabaseOperator;
+import azkaban.db.SQLTransaction;
+import azkaban.imagemgmt.exeception.ImageMgmtDaoException;
+import azkaban.imagemgmt.exeception.ImageMgmtException;
+import azkaban.imagemgmt.models.ImageRampup;
+import azkaban.imagemgmt.models.ImageRampup.StabilityTag;
+import azkaban.imagemgmt.models.ImageRampupPlan;
+import azkaban.imagemgmt.models.ImageRampupPlanRequest;
+import azkaban.imagemgmt.models.ImageRampupRequest;
+import azkaban.imagemgmt.models.ImageType;
+import azkaban.imagemgmt.models.ImageVersion;
+import azkaban.imagemgmt.models.ImageVersion.State;
+import com.google.common.collect.Iterables;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.commons.dbutils.ResultSetHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * DAO Implementation for accessing image rampup plan and rampups for an image type. This class
+ * contains implementation for methods such as create, get, update etc. for image rampups.
+ */
+@Singleton
+public class ImageRampupDaoImpl implements ImageRampupDao {
+
+  private static final Logger log = LoggerFactory.getLogger(ImageRampupDaoImpl.class);
+
+  private final DatabaseOperator databaseOperator;
+
+  private final ImageTypeDao imageTypeDao;
+
+  private final ImageVersionDao imageVersionDao;
+
+  private static String INSERT_IMAGE_RAMPUP_PLAN_QUERY = "insert into image_rampup_plan ( name, "
+      + "description, type_id, active, created_by, created_on, modified_by, modified_on) "
+      + "values (?, ?, ?, ?, ?, ?, ?, ?)";
+  private static String INSERT_IMAGE_RAMPUP_QUERY = "insert into image_rampup "
+      + "( plan_id, version_id, rampup_percentage, stability_tag, created_by, created_on, "
+      + "modified_by, modified_on) values (?, ?, ?, ?, ?, ?, ?, ?)";
+  private static String SELECT_IMAGE_RAMPUP_ACTIVE_PLAN_QUERY = "select irp.id, irp.name, "
+      + "irp.description, irp.active, it.name image_type_name, irp.created_on, irp.created_by, "
+      + "irp.modified_on, irp.modified_by from image_rampup_plan irp, image_types it where "
+      + "irp.type_id = it.id and it.name = ? and irp.active = ? ";
```

Comment:
There will be only one active rampup plan and corresponding check is already in place. If we allow more that one active rampup plan it will be messy and the version selection won't be proper. So we need to always ensure there is only one active plan. A new plan can be created or an existing active plan for an image type can be updated. Yeah we can add order by id desc just to be double sure there is always one active plan.

Reply:
In that case, order by irp.id, it.type_id would be good.

Reply:
As we are concern about only the active plan (joining with image_types table is just to get the type), so added "order by irp.id desc limit 1". Verified and this should work properly.